Membership Application: use state machine (aasm gem)

weedySeaDragon opened the pull request:
PT Story:  **Refactor to state machine (big one)**
 https://www.pivotaltracker.com/story/show/134267809

Use the 'aasm' (act as state machine) gem to manage the states, events, and transitions with the MembershipApplication.

### Changes proposed in this pull request:
1.  change 'status' to 'state'
2. use `aasm` gem ('acts as state machine'); put state machine logic into the model (move out of controller)
3. admin uses buttons to change the state of an application (instead of drop-down list). see screenshots (needed to add `.scss` styling for disabled buttons and form (input) buttons)
4. define acceptable states and transitions
5. add more detailed specifications and features relating to the membership application state
6. fixed Swedish wording
7. Proposed a rake task for populating the 'state' column for existing applications
8. Implement the new states and transitions discussed with Sussana, Ulrika, Pat, & Ashley
9. Add a checkbox on the user's edit application page.  If this checkbox is checked when they click 'submit,' the application will change state (from `waiting_for_applicant` to `ready_for_review`)
10. Added some RSpec matchers  to aasm


**State diagram and state transition table for MembershipApplication after applying this PR:**

![membership_application](https://cloud.githubusercontent.com/assets/673794/21972692/20c3d562-db70-11e6-83de-eac1cd27cb00.png)

This file is `./docs/membership_application.png`
(This was automatically generated by the aasm_statecharts gem [the version in https://github.com/weedySeaDragon/aasm_statecharts].  Hence the layout isn't the best. )


### Screenshots:

![screenshot-aasmstate-1-adminlanding](https://cloud.githubusercontent.com/assets/673794/21579418/ed2a4d76-cf62-11e6-9c8b-0de58ce77692.png)

![screenshot-aasmstate-1-admin-membershipstates](https://cloud.githubusercontent.com/assets/673794/21579419/ed2b4db6-cf62-11e6-8f7e-8653861419bc.png)

![screenshot-aasmstate-1-admin-membershipstates-en](https://cloud.githubusercontent.com/assets/673794/21579417/ed28f5f2-cf62-11e6-9fa9-ca43788fc6e2.png)

![screenshot-aasmstate-1-admin-needsinfo-en](https://cloud.githubusercontent.com/assets/673794/21579420/ed30dfba-cf62-11e6-86d3-a164aed44146.png)

#### Checkbox the users sees when she's on the edit form, IF it has been in the `waiting for applicant` state:
(Yes, the label and checkbox need to all be inline.  That can happen in another PR)
<img width="741" alt="screenshot-user-editapp-readyforreviewcheckbox-sv" src="https://cloud.githubusercontent.com/assets/673794/21972790/0a9858ca-db71-11e6-9857-e20ac589b850.png">


<img width="747" alt="screenshot-user-editapp-readyforreviewcheckbox" src="https://cloud.githubusercontent.com/assets/673794/21972789/0a954158-db71-11e6-89ff-bc69076b6987.png">

Ready for review:
@thesuss @patmbolger

weedySeaDragon committed:
add gem: aasm  (acts as state machine)

weedySeaDragon committed:
migration: add 'state' to MembershipApplication; db:migrate

weedySeaDragon committed:
MembershipApplication: add 'state'; add states and transitions; accept and reject actions; helper methods

weedySeaDragon committed:
update specifications to include 'state', events, and transitions (and not include 'status' anymore)

weedySeaDragon committed:
add entries for the different states of the MembershipApplication

weedySeaDragon committed:
routes for MembershipApplication

weedySeaDragon committed:
use 'state' to check if membership application is approved/accepted; assign the state provided in the hash (background, etc.)

weedySeaDragon committed:
added steps: number of times text should appear on a page; check the membership status line;

weedySeaDragon committed:
add styles for change membership application state buttons

weedySeaDragon committed:
change 'status' to 'state'

weedySeaDragon committed:
policy: change status to state; add actions information, accept, reject, need_info, cancel_need_info

weedySeaDragon committed:
add actions for changing membership state: accept, reject, need_info, cancel_need_info; move logic to model

weedySeaDragon committed:
(wip) features for membership_application state.  Still need to do sad paths

weedySeaDragon committed:
membership application state change: use buttons to change state; show the status before the buttons to change it; 'status' -> 'state'

weedySeaDragon committed:
diagram for MembershipApplication states and transformations

weedySeaDragon committed:
Merge branch 'develop' into sprint4-PT134267809--state-machine-aasm

thesuss commented on app/assets/stylesheets/custom-membership-application-state.scss:
```diff
@@ -0,0 +1,35 @@
+@import 'bootstrap/variables';
+@import 'bootstrap/mixins';
+@import 'bootstrap/buttons';
```
ignore this one

thesuss commented on app/assets/stylesheets/custom-membership-application-state.scss:
```diff
@@ -0,0 +1,35 @@
+@import 'bootstrap/variables';
+@import 'bootstrap/mixins';
```
ignore this one

weedySeaDragon committed:
style and formatting updates

weedySeaDragon committed:
add trailing newline to end of file

weedySeaDragon committed:
Merge remote-tracking branch 'origin/sprint4-PT134267809--state-machine-aasm' into sprint4-PT134267809--state-machine-aasm

weedySeaDragon committed:
change to :disabled (instead of selector); put each  selector in a comma sequence on its own single line

weedySeaDragon committed:
re-order cursor & opacity to make houndci-bot happy

weedySeaDragon committed:
re-order font-weight & display to make houndci-bot happy

thesuss commented:
Just a note that we have to fix the language before pressing the button on this one.

patmbolger commented on app/models/membership_application.rb:
```diff
@@ -9,17 +9,115 @@ class MembershipApplication < ApplicationRecord
+      transitions from: [:accepted, :rejected, :waiting_for_applicant, :pending], to: :pending, guard: :not_a_member?
+    end
+
+  end
```
Please add this line to app/controllers/application_controller.rb:
```
require 'aasm'
```
That will make the module available in rails console.  Without that, I get this error:
```
command: MembershipApplication.all.count
response:  NameError: uninitialized constant MembershipApplication::AASM
```

weedySeaDragon replied:
@patmbolger  I am able to get a correct response without adding that line.  Is it possible you need to run `bundle install` ?  And if you did run bundle install and are still getting the error, then perhaps Susanna can check to see what she gets.

patmbolger replied:
@weedySeaDragon - you're right .... I must have forgotten to run ```bundle```

weedySeaDragon replied:
whew.  glad it was that simple.  Otherwise I was afraid there was something nasty we'd have to untangle. :-)

patmbolger commented on app/models/membership_application.rb:
```diff
@@ -9,17 +9,115 @@ class MembershipApplication < ApplicationRecord
+
+  def accept_membership
+    begin
+      user.update({is_member: true})
```
Don't need braces if last (or sole) argument to a method is a hash, hence:
```
user.update(is_member: true)
```
reads clearer.

patmbolger commented on app/controllers/membership_applications_controller.rb:
```diff
@@ -45,11 +45,7 @@ def update
-        redirect_to edit_membership_application_url(@membership_application)
-        return
-      end
+      @membership_application.applicant_updated_info! if (current_user.id == @membership_application.user.id)
```
If the user (applicant) is 

1) editing the application, 
2) and the application is in state `:accepted`, `:rejected`, or `:waiting_for_applicant`
3) the user presses the "Save" button without actually changing anything (did not change any attributes nor upload any files)

then, the event logic:
```
event :applicant_updated_info do
  transitions from: [:accepted, :rejected, :waiting_for_applicant, :pending], to: :pending, guard: :not_a_member?
end
```
will cause a state change to `pending`, even though the user changed nothing.  

In this case, I would think the application should stay in its original state.

If that is the case, this action logic should detect that there are no changes and *not* fire a state change event.

thesuss replied:
Well, wether or not a user CAN/SHOULD edit their application is up for debate. (trying to set up a meeting tomorrow)

weedySeaDragon replied:
agree with both: shouldn't change the state if no changes were actually made, and we need to know if a user can edit their application (probably), and if a member can edit their application (don't know).

weedySeaDragon replied:
Per the meeting we had earlier today (2 January) with Ulrika, we decided that a user must explicitly mark the application as ready for review again; that will change the state to 'Ready for review' (the state formerly known as 'pending').  Just `saving` or even updating information by the user is not enough to change the status.

thesuss replied:
We also have to test that if a user edits their application and adds a category, submits it for review, the new category is not public until approved... 
I don't know if our system caters for this?

weedySeaDragon replied:
We don't do this yet.  Sounds like a user story & feature that needs to be written up. :-)

patmbolger commented on app/models/membership_application.rb:
```diff
@@ -9,17 +9,115 @@ class MembershipApplication < ApplicationRecord
+        company = Company.create!(company_number: company_number,
+                                  email: contact_email)
+      end
+      update(company: company)
```
I think that we need to update the `company` for the application *only if* we created a new company(?)

In that case, this statement should be in the ```unless``` block.

weedySeaDragon replied:
Actually, once I thought about this more (thanks for making me do that),  that's not quite right. 

The update line assigns the company instance (whether it was created or already existed) to the membership application.  Until an application is accepted, no company is assigned to that attribute. So that line needs to stay as is.

However, I did improve this by using `find_or_create_by!(company_number: company_number)` instead of the `unless` block.

patmbolger replied:
I did not know that (company not assigned to app until app is accepted).

What is the business rationale for that?

patmbolger replied:
@weedySeaDragon - I suggest adding a comment in this method to reflect this business rule that a company is not associated with a membership application until that application is accepted.  Otherwise, someone else reading the code might stumble over the same misunderstanding that I had.

weedySeaDragon replied:
I agree.  In the past, Thomas O. didn't approve of me adding comments to code, so I abstained.  (I didn't understand his boundaries for when it was or wasn't ok.)
I think comments in the code about this are good, but I also think this points to a need for documenting this outside of the code -- where is the right place to do this?  Somewhere on the project page on the AV site?  I think we -- the project as a whole -- need to decide on this.

(edited to add):  Specifications (features & rspecs) are a big part of this; we haven't been using them this way, though.  Should we?

patmbolger commented on app/views/membership_applications/_application_status_form.html.haml:
```diff
@@ -1,9 +1,15 @@
+
+  = button_to t('membership_applications.ask_applicant_for_info'), need_info_membership_application_path(@membership_application), {class: 'btn need-info', disabled: !(@membership_application.may_ask_applicant_for_info?)}
+
+  = button_to t('membership_applications.cancel_waiting_for_applicant'), cancel_need_info_membership_application_path(@membership_application), {class: 'btn cancel-need-info', disabled: !(@membership_application.may_cancel_waiting_for_applicant?)}
```
The event logic for above two buttons `accept` and `ask applicant for info` results in those two buttons being disabled when the admin is viewing a rejected application:
<img width="617" alt="screen shot 2017-01-01 at 9 35 53 am" src="https://cloud.githubusercontent.com/assets/9968213/21581592/b911ec12-d005-11e6-9cc8-a9744da4f220.png">

Yet the state change diagram shows that state change from `rejected` to either of `accepted` or `waiting for applicant` are allowed changes.

weedySeaDragon replied:
We haven't decided on the states and transitions yet.  I started with these, but need to change them according to decisions/feedback.

The state can change from `rejected` but only if `not_a_member?` is true.  (In the diagram, the guard conditions for a transition are in square brackets `[]`.)

patmbolger commented on features/update_membership_application.feature:
```diff
@@ -8,31 +8,282 @@ Feature: As an Admin
+  Scenario: Things go wrong when trying to reject an application
+
+
+  Scenario: Anna edits her application to provide needed info (from waiting for into to pending)
```
" .... (from waiting for `info` to pending)"

patmbolger commented on features/update_membership_application.feature:
```diff
@@ -8,31 +8,282 @@ Feature: As an Admin
+
+
+  # From needs info to...
+  Scenario: Admin changed from needs info to 'info is no longer needed (cancelled waiting for info)'
```
Might want to change ```needs info``` to ```waiting for applicant```, since that is the actual state name (makes it easier to follow the code and map to state changes).

patmbolger commented on features/update_membership_application.feature:
```diff
@@ -8,31 +8,282 @@ Feature: As an Admin
+    Then I should not see button t("membership_applications.waiting_for_applicant")
+
+
+  Scenario: Admin cannot change from rejected to accepted
```
Should say:

Admin *can* change from rejected to accepted

patmbolger commented on features/update_membership_application.feature:
```diff
@@ -8,31 +8,282 @@ Feature: As an Admin
 
+  Scenario: Admin cannot change from rejected to cancel needs info
+    Given I am on "LarsRejected" application page
+    Then I should not see button t("membership_applications.cancel_waiting_for_applicant")
```
I tried to map state changes as shown in the STD to this feature.  In this snapshot, yellow highlights the state changes I did not see in the feature.  (those may be present in other feature tests - I did not look - however, it would be good to have all application state changes tested in this feature).

<img width="597" alt="screen shot 2017-01-01 at 12 07 14 pm" src="https://cloud.githubusercontent.com/assets/9968213/21582156/e19de612-d01a-11e6-964f-2896adc0ecdd.png">

patmbolger reviewed (commented):
Great work on the PR .... I did not check everything as the PR is quite large - will check more after next iteration.

Please see comments.

weedySeaDragon committed:
don't need braces if last or only arg is a hash

weedySeaDragon committed:
find or create the company based on company_number (for an accepted membership application)

weedySeaDragon committed:
rename 'pending' to 'under_review'

weedySeaDragon committed:
if rejected, can only view an application -- cannot edit it

weedySeaDragon committed:
(wip) add scenarios for 'new' state; add checks for 'ready for review' and 'started review'; use 'waiting for applicant'; re-organize

weedySeaDragon committed:
cannot transition to accept or reject if state = waiting for applicant

patmbolger reviewed (commented):
Thanks for comments so far ....

thesuss commented:
Probably so not to show the potentially added categories on the company until they are indeed added?

weedySeaDragon commented:
rationale for not having a fully instantiated company until accepted:

Part of the review process is to verify the  Organization number (company number).  (Done mostly by the gem `orgnummer` but also by the SHF admin, I think?)

We only need and use the company information once the applicant has been approved:  visitors can only see companies on the site if they are approved/accepted companies.

If the applicant is not approved/accepted, then there seems no point in instantiating a company and then, when they're rejected, deleting it.  

Make sense, @patmbolger  and @thesuss ?

weedySeaDragon committed:
fix step: was looking up the company, not the application (copy and paste error?)

(cherry picked from commit 4c92b54)

weedySeaDragon committed:
db: default state should be 'under_review' (was: 'Pending')

weedySeaDragon committed:
change 'pending' to 'under_review'

weedySeaDragon committed:
change to pending; must be in 'waiting for applicant' state for user to be able to edit

weedySeaDragon committed:
updated diagram

weedySeaDragon commented:
I know that coverage has declined -- it's because I don't have tests that capture the exceptions raised.  I wanted to go ahead and try to get this wrapped up.  We'll be revising and adding to the states soon and can make sure the coverage increases again.

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+   # accept.success
+    accept:
+      success: Medlemsansökan har godkänts.
+      error: Det uppstod ett fel när man försöker att godkänna medlemskap.
```
Det uppstod ett fel vid godkännande av medlemskap.

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
 
+    reject:
+      success: Den ansökan har avslagits.
+      error: Det uppstod ett fel när man försöker att avslå ansökan.
```
Det uppstod ett fel vid avslag av medlemskap.

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+      error: Det uppstod ett fel när man försöker att godkänna medlemskap.
 
+    reject:
+      success: Den ansökan har avslagits.
```
Medlemsansökan har avslagits.

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+      error: Det uppstod ett fel när man försöker att avslå ansökan.
+
+    need_info:
+      success: 'Ansökan har markerats som behöver information. (Skicka e-post till den sökande.)'
```
Ansökan har markerats som: "Behöver kompletteras". (Skicka e-post till den sökande.)

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+
+    need_info:
+      success: 'Ansökan har markerats som behöver information. (Skicka e-post till den sökande.)'
+      error: "Det uppstod ett fel när man försöker ställa in ansökan 'behöver information.'"
```
Det uppstod ett fel när systemet försökte sätta status: "Behöver kompletteras".

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+
+    cancel_need_info:
+      success: Ansökan har skett förändringar inte längre behöver information från sökanden.
+      error: "Det uppstod ett fel när man försöker ställa in ansökan till 'inte längre behöver information från den sökande.'"
```
Det uppstod ett fel när systemet försökte ångra status: "Behöver kompletteras".

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+      error: "Det uppstod ett fel när man försöker ställa in ansökan 'behöver information.'"
+
+    cancel_need_info:
+      success: Ansökan har skett förändringar inte längre behöver information från sökanden.
```
Ansökan har ändrats och behöver ej längre kompletteras av sökande.

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+      error: "Det uppstod ett fel när man försöker ställa in ansökan till 'inte längre behöver information från den sökande.'"
+
+
+    accept: &event_accept Godkänna
```
Godkänn

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+
+    accept: &event_accept Godkänna
     accepted: &status_accepted Godkänd
+    reject: &event_reject Avböja
```
Avböj

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
-    pending: &status_pending Inlämnad
+    under_review: &status_under_review Behandlas
     pending_completion: &status_pending_completion Väntar på sökande
+    ask_applicant_for_info: &event_ask_applicant_for_info Be Sökande Info
```
Behöver kompletteras

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+    under_review: &status_under_review Behandlas
     pending_completion: &status_pending_completion Väntar på sökande
+    ask_applicant_for_info: &event_ask_applicant_for_info Be Sökande Info
+    cancel_waiting_for_applicant: &event_cancel_waiting_for_applicant Avbryt Väntar på Sökande Info
```
Ångra Behöver kompletteras

thesuss commented on config/locales/sv.yml:
```diff
@@ -179,11 +183,35 @@ sv:
+    ask_applicant_for_info: &event_ask_applicant_for_info Be Sökande Info
+    cancel_waiting_for_applicant: &event_cancel_waiting_for_applicant Avbryt Väntar på Sökande Info
+    waiting_for_applicant: &status_waiting_for_applicant  Väntar på sökande
+    applicant_updated_info: &event_applicant_updated_info Sökanden uppdaterad information
```
Sökanden har kompletterat sin ansökan

thesuss commented on config/locales/en.yml:
```diff
@@ -67,6 +67,7 @@ en:
         last_name: &last_name Last name
         contact_email: &contact_email Email
         status: &status Status
```
Used anywhere?

weedySeaDragon replied:
Nope, not anymore.  Good catch.

weedySeaDragon commented:
This PR is almost finished (whew).  The 1 remaining thing to be done is to initialize the 'state' column  for existing records. 

I've written a rake task that does this, but haven't committed it yet.   Would this be the best way to handle it, since it's a 1 time thing?  

(I'll commit the rake task in a bit so that it can be reviewed.)

weedySeaDragon committed:
rake task to set the value of the 'state' column based on the 'status' column

weedySeaDragon committed:
fix translations

weedySeaDragon committed:
remove 'status' entries ('state' replaced it)

weedySeaDragon committed:
add transition table to the state transition diagram

weedySeaDragon committed:
change any existing 'Pending' states to 'under_review' during migration

patmbolger commented:
Yesterday we had talked about how the application transitions from "waiting for applicant" to "ready for another review" state, once the applicant has finished updating their info.  (that is, we don't transition immediately back to "under review")

However, the state transition table shows a transition to "under review" state:

<img width="600" alt="screen shot 2017-01-12 at 3 42 15 pm" src="https://cloud.githubusercontent.com/assets/9968213/21907570/06995ea8-d8de-11e6-9183-bbdfa6247996.png">

Did I get that wrong?  If not, perhaps we need to remove this particular transition from the PR, or add the state "ready for another review" to accommodate this transition.

weedySeaDragon commented:
@patmbolger You're right:  --  _where we want to end up_ is with a `ready for review` state.
But this PR isn't there.  We just have the `under_review` state.  (which used to be called "pending")

What this really signals to me is that this PR should either not change any of the states that we had originally, or  else go ahead and implement the new states, i.e. _where we want to end up_ including the `new`, `ready_for_review` states, etc.

My inclination is to go ahead and implement the new states.  Going backwards to where the states were before this PR started would be just as "big" at this point.

Your thoughts?

patmbolger commented:
I agree .... might as well implement the missing two states as well since you're mostly there anyway.

weedySeaDragon committed:
update: owners can change the state of an application; improve readability: add text title to examples; give users descriptive emails

weedySeaDragon committed:
update for state attribute (owner can now edit it); improve readability by renaming, reorganizing

weedySeaDragon committed:
rename method from 'edit_state?' to 'can_edit_state?' to better express intent

weedySeaDragon committed:
is an owner if the record is an instance of a MembershipApplication (vs. the class) [and the other existing checks]

weedySeaDragon committed:
add my additional RSpec matchers for aasm

weedySeaDragon committed:
MembershipApplication factory initial state should be set to :new

weedySeaDragon committed:
avoid name collision with blank? method so that these matchers can be used outside of Rails

weedySeaDragon committed:
fix syntax (copy and paste) error

weedySeaDragon committed:
should check all valid aasm events on the class (not instance)

weedySeaDragon committed:
cleaned up specs: simplified with shared examples, aliased for readability, use aasm matchers I wrote

weedySeaDragon committed:
allow the checkbox 'marked_ready_for_review'

weedySeaDragon committed:
ready for review checkbox

weedySeaDragon committed:
add is ready for review checkbox if the application is in the right state

weedySeaDragon committed:
add 'start reivew' button

weedySeaDragon committed:
initial state for an application should now be 'new'

weedySeaDragon committed:
add steps for checkboxes

weedySeaDragon committed:
transition table and state diagram for MembershipApplication

generated with aasm_statecharts -a -t -d ./docs
using the weedySeaDragon/aasm_statecharts gem from github

weedySeaDragon committed:
update features to reflect new and additional states

weedySeaDragon committed:
add start-review action to the routes

weedySeaDragon committed:
only an admin can see the change status buttons

weedySeaDragon committed:
add entries for ready for review

weedySeaDragon committed:
add action for start review; add check to see if the user marked the form 'ready to be reviewed'

weedySeaDragon committed:
add instance var to track if it's marked as ready for review (used by the edit form checkbox); update state transitions

weedySeaDragon committed:
update gemfile.lock (not sure we should really be checking this in, though!)

weedySeaDragon commented on spec/support/ae_aasm_matchers/have_valid_state.rb:
```diff
@@ -0,0 +1,32 @@
+    "expected :#{expected} to be a valid state (for_state_machine :#{@state_machine_name})"
+  end
+
+  failure_message_when_negated do |obj|
```
do need to have it there so that users can pass in that parameter.

weedySeaDragon commented on spec/support/ae_aasm_matchers/have_valid_transition_from_to.rb:
```diff
@@ -0,0 +1,31 @@
+    "#{expected} is a valid state (for_state_machine :#{@state_machine_name})"
+  end
+
+  failure_message do |obj|
```
do need to have it there so that users can pass in that parameter.

weedySeaDragon commented on spec/support/ae_aasm_matchers/have_valid_transition_from_to.rb:
```diff
@@ -0,0 +1,31 @@
+    "expected :#{expected} to be a valid state (for_state_machine :#{@state_machine_name})"
+  end
+
+  failure_message_when_negated do |obj|
```
do need to have it there so that users can pass in that parameter.

weedySeaDragon commented on spec/support/ae_aasm_matchers/have_valid_state.rb:
```diff
@@ -0,0 +1,32 @@
+    "#{expected} is a valid state (for_state_machine :#{@state_machine_name})"
+  end
+
+  failure_message do |obj|
```
do need to have it there so that users can pass in that parameter.

weedySeaDragon commented on spec/support/ae_aasm_matchers/have_valid_event.rb:
```diff
@@ -0,0 +1,34 @@
+    "expected that :#{expected} is a valid event (for_state_machine :#{@state_machine_name})"
+  end
+
+  failure_message_when_negated do |obj|
```
do need to have it there so that users can pass in that parameter.

weedySeaDragon commented on spec/support/ae_aasm_matchers/have_valid_event.rb:
```diff
@@ -0,0 +1,34 @@
+    "#{expected} is a valid event (for_state_machine :#{@state_machine_name})"
+  end
+
+  failure_message do |obj|
```
do need to have it there so that users can pass in that parameter.

weedySeaDragon committed:
added trailing newline

weedySeaDragon committed:
Merge branch 'develop' into sprint4-PT134267809--state-machine-aasm

weedySeaDragon committed:
fix status -> state, accepted.  (also no longer need the is_member column for creating users)
